lms/djangoapps/mobile_api/tests/test_middleware.py

```diff
@@ -0,0 +1,154 @@
+"""
+Tests for Version Based App Upgrade Middleware
+"""
+from datetime import datetime
+import ddt
+from django.core.cache import cache, caches
+from django.http import HttpRequest, HttpResponse
+from django.test import TestCase
+import mock
+from pytz import UTC
+from mobile_api.middleware import AppVersionUpgrade
+from mobile_api.models import AppVersionConfig
+
+
+@ddt.ddt
+class TestAppVersionUpgradeMiddleware(TestCase):
+    """
+    Tests for version based app upgrade middleware
+    """
+    def setUp(self):
+        super(TestAppVersionUpgradeMiddleware, self).setUp()
+        self.middleware = AppVersionUpgrade()
+        self.set_app_version_config()
+        cache.clear()
+
+    def set_app_version_config(self):
+        """ Creates configuration data for platform versions """
+        AppVersionConfig(platform="iOS", version="1.1.1", expire_at=None, enabled=True).save()
+        AppVersionConfig(
+            platform="iOS",
+            version="2.2.2",
+            expire_at=datetime(2014, 01, 01, tzinfo=UTC),
+            enabled=True
+        ).save()
+        AppVersionConfig(
+            platform="iOS",
+            version="4.4.4",
+            expire_at=datetime(9000, 01, 01, tzinfo=UTC),
+            enabled=True
+        ).save()
+        AppVersionConfig(platform="iOS", version="6.6.6", expire_at=None, enabled=True).save()
+
+        AppVersionConfig(platform="Android", version="1.1.1", expire_at=None, enabled=True).save()
+        AppVersionConfig(
+            platform="Android",
+            version="2.2.2",
+            expire_at=datetime(2014, 01, 01, tzinfo=UTC),
+            enabled=True
+        ).save()
+        AppVersionConfig(
+            platform="Android",
+            version="4.4.4",
+            expire_at=datetime(5000, 01, 01, tzinfo=UTC),
+            enabled=True
+        ).save()
+        AppVersionConfig(platform="Android", version="8.8.8", expire_at=None, enabled=True).save()
+
+    def process_middleware(self, user_agent):
+        """ Helper function that makes calls to middle process_request and process_response """
+        fake_request = HttpRequest()
+        fake_request.META['HTTP_USER_AGENT'] = user_agent
+        request_response = self.middleware.process_request(fake_request)
```

Comment:
Check that process_request() calls `get_many()` once. This insures that the mocking we are doing below is working correctly (so for example, if we are not mocking correctly, the call_count would be zero irrespective of the number of calls to the method).